db/db_impl/db_impl.h

```diff
@@ -1608,7 +1608,8 @@ class DBImpl : public DB {
   // state needs flush or compaction.
   void InstallSuperVersionAndScheduleWork(
       ColumnFamilyData* cfd, SuperVersionContext* sv_context,
-      const MutableCFOptions& mutable_cf_options);
+      const MutableCFOptions& mutable_cf_options,
+      std::list<std::unique_ptr<FlushJobInfo>>* flush_jobs_info);
```

Comment:
I *suspect* that at least in theory, we could have the same problem with nothing but compactions, e.g. compaction A creates SST 11, then schedules compaction B that trivially moves SST 11 to a lower level. Because `InstallSuperVersionAndScheduleWork` precedes `NotifyOnCompactionCompleted`, I think in this case the client could get notified about compaction B before A. If so, we would have to make this more general and pass in some sort of callback.

Reply:
you are correct, disordered notifications of compaction completed are still possible. let me see if I can make it more general.

Reply:
One way would be to pass in an optional function object that encapsulates the notification.

Reply:
Yes, that's how I did it, but I found some other issues, still looking at it.